File: com/white/LargestRectangleInHistogram.java
package com.white;

import java.util.Stack;

public class LargestRectangleInHistogram {

	// 84. Largest Rectangle in Histogram
	
	public static int largestRectangleArea(int[] heights) {
        
		Stack<Integer> stack = new Stack<>();
		int maxArea = 0;
		int i = 0; 
		
		while (i<heights.length) {
			
			if (stack.empty() || heights[stack.peek()]<heights[i] ) {
				stack.push(i);
				i++;
			} else {
				int currMax = stack.pop();
				int area = heights[currMax] * (stack.empty() ? i : (i-1-stack.peek()));
				
				if (area>maxArea) {
					maxArea = area;
				}
			}
		}
		
		while (!stack.empty()) {
			int currMax = stack.pop();
			int area = heights[currMax] * (stack.empty() ? i : (i-1-stack.peek()));
			
			if (area>maxArea) {
				maxArea = area;
			}
		}
		
		return maxArea;
    }
}

File: LeetcodeCore/LargestRectangleInHistogram.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class LargestRectangleInHistogram
    {
        // 84. Largest Rectangle in Histogram
        // Monotonic stack idea is hard to grasp
        public int LargestRectangleArea(int[] heights)
        {
            var stack = new Stack<int>();
            var maxArea = 0;
            var i = 0;

            while (i < heights.Length)
            {
                if (stack.Count == 0 || heights[i] > heights[stack.Peek()])
                {
                    stack.Push(i);
                    i++;
                }
                else
                {
                    var currMax = stack.Pop();
                    var currArea = heights[currMax] * (stack.Count == 0 ? i : (i - 1 - stack.Peek()));
                    maxArea = Math.Max(maxArea, currArea);
                }
            }

            while (stack.Count > 0)
            {
                var currMax = stack.Pop();
                var currArea = heights[currMax] * (stack.Count == 0 ? i : (i - 1 - stack.Peek()));
                maxArea = Math.Max(maxArea, currArea);
            }

            return maxArea;
        }
    }
}
